README.md
## aframe-tangram-terrain-component

A [Tangram](https://github.com/tangrams/tangram) terrain component for [A-Frame](https://aframe.io) with batteries included. 

> Supports A-Frame 0.8.0.

![Example](doc/example.jpg)

### API

#### `tangram-terrain` component

This component obtains a normalmap from the [Amazon public dataset S3](https://aws.amazon.com/public-datasets/terrain/) servers and applies a overlay texture according the `style` specification. The normlmap is used both as a `displacement` map and if wanted also a a `normalmap`. By using the _GPU picking technique_ everything is done on the shader for height calculation. Renders a depth buffer to an external _scene_, so that the height from the normal map alpha channel is read by the GPU. Offers Level-of-Detail (LOD) using buffer geometry draw ranges.

##### Schema
| Property | Description | Default Value |
| -------- | ----------- | ------------- |
| apiKey | Sets the global API key of the overlay map. May be empty if the style needs no API. | "" |
| style | The style definition document for the ovleray style. Must point to a custom style or to a [basemap style](https://www.nextzen.org/). May need an API key set. | "" |
| includeOceans | Include Ocean heightmap data between. To be implemented. | false |
| center | Center of the map, in the form of [longitude, latitude] | [0, 0] |
| zoom | The zoom level of the map. | 13 |
| pxToWorldRatio | The multiplication factor between meters in A-Frame and the pixels of the map. ie; when set to 100, will display 100 pixels per 1 meter in world space. (see [a note on fidelity](#a-note-on-fidelity)) | 100 |
| lodCount | Number of level of details for _geometry_. Set to 1 of no LOD is needed. | 1 |
| lod | Level-Of-Detail in the range between 1 and _lodCount_. May be set during runtime. | 1 |
| vertexNormals | Decides if a normalmap shall be applied to the material. Needed for lighting. Due to smart terrain tiles no additional data needs to transfered, just you are GPU needs to work harder. | true |

##### Events
| Name | Data | Description |
| -------- | ----------- | ------------- |
| tangram-terrain-loaded | None| Fired when the terrain that means the elevation data and the overlay map has finished loading or the corresponding Level-Of-Detail has been loaded. Only by then the public API will be usable. |



##### API
| Name | Data | Description |
| -------- | ----------- | ------------- |
| project | _lon_, _lat_| Returns the pixel x and y and z (depth) coordinates of the given longitude and latitude. |
| unproject | _x_, _y_| Gives the longitude and latitude of the world coordinates. |
| unprojectHeight | _x_, _y_| Gives the depth value (GPU picked from the heightmap) from the pixel coordinates, only if the `depthBuffer` value is set to _true_. Value is scaled according to the `displacementScale` scale and added the `displacementBias` value. To obtain real height values |
| unprojectHeightInMeters | _x_, _y_| Gives the height value in meters. No exact values are possible though. Range between 0 and 8900 (Mount Everest) |
| renderDepthBuffer | | Renders the heightmap depth buffer. Needs to be only called manually if the heigtmap is changed programmtically |


#### `tangram-static-terrain` component

This component works offline by specifying the map and normalmap by the user. To take the normalmap, press `<ctrl> + <alt> + t` on the keyboard. To take the styled map, press `<ctrl> + <alt> + <shift> + t` on the keyboard.

##### Schema
| Property | Description | Default Value |
| -------- | ----------- | ------------- |
| map | Reference to the visible texture. | "" |
| normalmap | Reference to the normalmap texture. | "" |
| bounds | String of four elements obtained from the console log. | "0, 0, 0, 0" |
| lodCount | See `tangram-terrain` component | 1 |
| lod | See `tangram-terrain` component | 1 |
| vertextNormals | See `tangram-terrain` component | false |

### Styling
The Tangram map is styled within a (set) of YAML files. See the [Tangram documentation](https://mapzen.com/documentation/tangram/) for details. 

### A note on fidelity

The higher `pxToWorldRatio`, the more map area will be displayed per world
unit. That canvas has to be translated into a plane in world space. This is
combined with the width and height in world space (from geometry.width and
geometry.height on the entity) to set up the plane for rendering in 3D.

The map is rendered as a texture on a 3D plane. For best performance, texture
sizes should be kept to powers of 2. Keeping this in mind, you should work to
ensure `width * pxToWorldRatio` and `height * pxToWorldRatio` are powers of 2.

### Dependencies
The Mapzen styling documents are in the YAML format, so you need a possiblity to require those files.
If you are using _Webpack_ install
`npm install yml-loader --save-dev` 
and configure the webpack configuration file

If you are using browserify install the
`nmp install yamlify --save-dev` 
and give pass the transform (-t) parameter to browserify.

### Installation

#### Browser

Install and use by directly including the [browser files](dist):

```html
<head>
  <title>My A-Frame Scene</title>
  <script src="https://aframe.io/releases/0.8.0/aframe.min.js"></script>
  <script src="https://unpkg.com/aframe-tangram-terrain-component/dist/aframe-tangram-terrain-component.min.js"></script>
</head>

<body>
  <a-scene>

    <a-assets>
      <a-asset-item id="yamlElevation" src="https://raw.githubusercontent.com/tangrams/terrain-demos/master/styles/terrarium-elevation.yaml" />
    </a-assets>

    <a-tangram-terrain
      map-style="#yamlElevation" 
      zoom="10" 
      center="15.8056, 7.7671" 
      px-world-ratio="12"
      wireframe="false" 
      height-scale="40"
      width="25" depth="25">
    </a-tangram-terrain>
  </a-scene>
</body>
```

<!-- If component is accepted to the Registry, uncomment this. -->
<!--
Or with [angle](https://npmjs.com/package/angle/), you can install the proper
version of the component straight into your HTML file, respective to your
version of A-Frame:

```sh
angle install aframe-tangram-terrain-component
```
-->

#### npm

Install via npm:

```bash
npm install aframe-tangram-terrain-component
```

Then require and use.

```js
require('aframe');
require('aframe-tangram-terrain-component');
```

#### Required step for Webpack
To get the import working you need to tell _Webpack_ not to parse the _Tangram_ library otherwise you will run into [this error](https://github.com/tangrams/tangram/issues/559). So you need to add this in your _webpack.config.js_.

```js
module: {
    noParse: [/tangram\/dist\/tangram/],
}
```

See also [here](https://github.com/tangrams/tangram-play/wiki/Advanced-Tangram-for-front-end-engineers:-bundlers,-frameworks,-etc) for more information.


### Known issues
* The maximum texture size (configured by the _pxToWorldRatio_ and _size_ parameters of the overlay map is restricted on mobile devies.  
* The _Tangram_ library only allows [one instance on a page] (https://github.com/tangrams/tangram/issues/350) that means a tiling algorithm cannot instance terrain entities in parallel! Hopefully this restriction of the _Tangram_ library may fall in the future.

### Implementation details
To save bandwith and increase speed the elevation data is backed into the [normal map of the tiling webservice](https://s3.amazonaws.com/elevation-tiles-prod/normal/{z}/{x}/{y}.png), which is all done in a shader by the _Tangram_ library. _Three.JS_ does not support normals with displacement alpha channels so I created a custom _MeshStandardImplementation_ that respects this format. 


CHANGELOG.md
=3.0.0=

=2.1.0=
* Various render optimizations (Level-Of-Detail)
* Material LOD
* Geometry LOD
* Lower heightmap resolutions


=2.0.0=
* Refactored Tangram Usage into singleton _system_
* Components keep state

=1.0.0=
* Initial Release

tangram-heightmap.js
/* global AFRAME */

const latLonFrom = require('./utils').latLonFrom
const leafletOptions = require('./utils').leafletOptions
const getCanvasContainerAssetElement = require('./utils').getCanvasContainerAssetElement
const processStyle = require('./utils').processStyle
const processCanvasElement = require('./utils').processCanvasElement

if (typeof AFRAME === 'undefined') {
    throw new Error('Component attempted to register before AFRAME was available.');
}

const cuid = require('cuid')

const heightmapStyle = require('./heightmap-style.yaml');
require('leaflet')

const HEIGHTMAP_LOADED_EVENT = 'tangram-heightmap-loaded';

// TODO make configurable?
var tempFactor = 1; // size of heightMapCanvas relative to main canvas: 1/n

AFRAME.registerComponent('tangram-heightmap', {
    dependencies: [
        'geometry',
        'material'
    ],

    schema: {
        mapzenAPIKey: {
            default: ''
        },
        style: {
            type: "asset",
            default: ''
        },
        scaleFactor: {
            type: 'int',
            default: 1
        },
        center: {
            // lat lon
            default: [0, 0], //[47.7671, 15.8056], // Schneeberg
            type: 'array',
        },
        /**
            [0] southwest
            [1] northeast
        */
        maxBounds: {
            default: undefined,
            type: 'array',
            parse: value => {
                return value
            },
        },
        fitBounds: {
            default: undefined, //GROSSGLOCKNER,
            type: 'array',
            parse: value => {
                return value
            },
        },
        zoom: {
            default: 13
        },
        wireframe: {
            type: 'boolean',
            default: false
        },
        pxToWorldRatio: {
            type: 'int',
            default: 10
        },
        canvasOffsetPx: {
            type: 'int',
            default: 9999 // debug
        },
        // set the highest altitude
        highestAltitudeMeter: {
            type: 'int',
            default: undefined
        }
    },

    multiple: false,

    init: function() {
        this._minHeight = 0
        this._maxHeight = 0
        this.analysed = false
        this.analysing = false
        this._mapInstance = null
        this._scene = null

        this.altitudeAddition = 0

        this.terrainData = []

        this._initHeightMap()
    },
    _initHeightMap: function() {

        var data = this.data

        const geomComponent = this.el.components.geometry;
        var width = geomComponent.data.segmentsWidth
        var height = geomComponent.data.segmentsHeight

        this._canvasContainerId = cuid();
        const canvasContainer = getCanvasContainerAssetElement(this._canvasContainerId,
            width, height, data.canvasOffsetPx);

        var map = L.map(canvasContainer, leafletOptions);

        var layer = Tangram.leafletLayer({
            scene: heightmapStyle,
            global: {
                sdk_mapzen_api_key: data.mapzenAPIKey
            },
            attribution: '',
            postUpdate: _ => {
                // three stages:
                // 1) start analysis
                if (!this.analysing && !this.analysed) {
                    //console.log("EXPOSING")
                    this._expose()
                }
                // 2) continue analysis
                else if (this.analysing && !this.analysed) {
                    //console.log("START")
                    this._start_analysis();
                    this.analysed = false
                }
                /*else if (this.analysed) {
                                       // reset after next update (however that might be done)
                                       //console.log("ANALYSED")
                                       this._createTerrain()
                                       this.analysed = false
                                   }*/
            }
        });

        var scene = this._scene = layer.scene

        layer.on('init', _ => {
            // resetViewComplete();
            scene.subscribe({
                // will be triggered when tiles are finished loading
                // and also manually by the moveend event
                view_complete: function() {}
            });
            this.scene_loaded = true;

            var heightMapCanvas = document.createElement("canvas")
            heightMapCanvas.width = scene.canvas.width / tempFactor
            heightMapCanvas.height = scene.canvas.height / tempFactor

            this.heightMapCanvas = heightMapCanvas
        });
        layer.addTo(map);
        this._mapInstance = map

        if (data.maxBounds) this._mapInstance.setMaxBounds(L.latLngBounds(this.data.maxBounds))
        if (data.fitBounds) this._mapInstance.fitBounds(L.latLngBounds(this.data.fitBounds))
        this._mapInstance.setView(latLonFrom(this.data.center), this.data.zoom)
    },
    _expose: function() {
        this.analysing = true;
        if (this.scene_loaded) {
            this._start_analysis();
        } else {
            // wait for scene to initialize first
            this.scene.initializing.then(function() {
                this._start_analysis();
            });
        }
    },
    _start_analysis: function() {

        // based on https://github.com/tangrams/heightmapper/blob/gh-pages/main.js
        var scene = this._scene
        var canvas = this.heightMapCanvas

        var ctx = canvas.getContext("2d");
        ctx.clearRect(0, 0, canvas.width, canvas.height);

        // redraw canvas smaller in testing canvas, for speed
        ctx.drawImage(scene.canvas, 0, 0, scene.canvas.width / tempFactor, scene.canvas.height / tempFactor);

        // get all the pixels
        var pixels = ctx.getImageData(0, 0, canvas.width, canvas.height);

        var val;
        var counts = {};
        var empty = true;
        var max = 0,
            min = 255;


        const geomComponent = this.el.components.geometry;
        // only check every 4th pixel (vary with browser size)
        // var stride = Math.round(img.height * img.width / 1000000);
        // 4 = only sample the red value in [R, G, B, A]
        for (var i = 0; i < canvas.height * canvas.width * 4; i += 4) {
            val = pixels.data[i];
            var alpha = pixels.data[i + 3];
            if (alpha === 0) { // empty pixel, skip to the next one
                continue;
            }
            // if we got this far, we found at least one non-empty pixel!
            empty = false;

            // update counts, to get a histogram
            counts[val] = counts[val] ? counts[val] + 1 : 1;

            // update min and max so far
            min = Math.min(min, val);
            max = Math.max(max, val);

            this.terrainData.push(val)
        }

        // range is 0 to 255 which is 8900 meters according to heightmap-style
        this._minHeight = min
        this._maxHeight = max

        var highestMeter = max / 255 * 8900

        if (this.data.highestAltitudeMeter) {
            this.altitudeAddition = this.data.highestAltitudeMeter - highestMeter
        }

        // TODO Schneeberg is 2076 m
        console.log(min + ' max ' + max)
        console.log(max / 255 * 8900)
        console.log((max / 255 * 8900) + this.altitudeAddition)
        console.log(min / 255 * 8900)

        if (empty) {
            // no pixels found, skip the analysis
            return false;
        }


        this.analysing = false;
        this.analysed = true
        this._createTerrain()
    },
    _initMap: function(geometry) {

        var data = this.data

        const geomComponent = this.el.components.geometry;
        var width = Math.min(4096, THREE.Math.nextPowerOfTwo(geomComponent.data.width * data.pxToWorldRatio))
        var height = Math.min(4096, THREE.Math.nextPowerOfTwo(geomComponent.data.height * data.pxToWorldRatio))

        console.log(geomComponent.data.width + ' ' + width)

        var _canvasContainerId = cuid();
        const canvasContainer = getCanvasContainerAssetElement(_canvasContainerId,
            width, height, data.canvasOffsetPx);


        const renderer = L.canvas({
            padding: 0
        })

        const options = Object.assign({
                renderer
            },
            leafletOptions)

        var map = L.map(canvasContainer, options);


        const sceneStyle = processStyle(data.style);

        var layer = Tangram.leafletLayer({
            scene: sceneStyle,
            global: {
                sdk_mapzen_api_key: data.mapzenAPIKey
            },
            attribution: '',
            postUpdate: _ => {

                /*
                processCanvasElement(canvasContainer)
                const canvasId = document.querySelector(`#${_canvasContainerId} canvas`).id;
                this.el.setAttribute('material', 'width', `1024`);
                this.el.setAttribute('material', 'height', `1024`);
                this.el.setAttribute('material', 'src', `#${canvasId}`);
                */

                var texture = new THREE.CanvasTexture(layer.scene.canvas);
                texture.wrapS = THREE.ClampToEdgeWrapping;
                texture.wrapT = THREE.ClampToEdgeWrapping;
                var mesh = this.el.getObject3D('mesh')
                mesh.material.map = texture
                mesh.material.needsUpdate = true


                var mesh = new THREE.Mesh(geometry,
                    new THREE.MeshBasicMaterial({
                        map: texture,
                        wireframe: this.data.wireframe,
                        transparent: false
                    }));
                this.el.setObject3D('mesh', mesh)

                var mesh = this.el.getOrCreateObject3D('mesh', THREE.Mesh);
                mesh.geometry = new THREE.Geometry();

                this.el.emit(HEIGHTMAP_LOADED_EVENT);
            }
        });
        layer.addTo(map);

        const heightMapBounds = this._mapInstance.getBounds()
        map.fitBounds(heightMapBounds)
    },
    _createTerrain: function() {

        if (!this.analysed) return

        var data = this.terrainData

        const {
            width: elWidth,
            height: elHeight,
            segmentsWidth: elSegmentsWidth,
            segmentsHeight: elSegmentsHeight
        } = this.el.components.geometry.data;

        var geometry = new THREE.PlaneBufferGeometry(
            elWidth, elHeight,
            elSegmentsWidth - 1, elSegmentsHeight - 1)
        var vertices = geometry.attributes.position.array;
        //https://stackoverflow.com/questions/37927031/how-to-update-the-topology-of-a-geometry-efficiently-in-threejs
        for (var i = 0, j = 0, l = vertices.length; i < l; i++, j += 3) {
            // only set z values (note: planes are not standing by default)
            vertices[j + 2] = this._scale(data[i] + this.altitudeAddition)
        }
        geometry.computeFaceNormals();
        geometry.computeBoundingBox();

        //var texture = new THREE.CanvasTexture(this.heightMapCanvas);
        this._initMap(geometry)
    },
    _scale: function(value) {

        const {
            width: elWidth,
            segmentsWidth: elSegmentsWidth,
        } = this.el.components.geometry.data;

        const densityFactor = elWidth / elSegmentsWidth
        const zoomScaleFactor = this.data.zoom * 0.5 //this._mapInstance.getZoom()

        var height = (value * 0.05) * zoomScaleFactor * densityFactor * this.data.scaleFactor;
        return height ? height - this._minHeight : 0
    },


    remove: function() {},

    tick: function(delta, time) {},

    _getAltitudeFromXY: function(x, y) {
        const idx = this._scene.canvas.width * y + x
        return this.terrainData[idx] / 255 * 8900 + this.altitudeAddition
    },

    project(lon, lat) {

        // The position (origin at top-left corner) in pixel space
        let {
            x: pxX,
            y: pxY
        } = this._mapInstance.latLngToLayerPoint([lat, lon]);

        const {
            width: elWidth,
            height: elHeight
        } = this.el.components.geometry.data;


        const idx = this._scene.canvas.width * pxY + pxX
        var z = this._scale(this.terrainData[idx] + this.altitudeAddition)

        pxX /= this._scene.canvas.width
        pxY /= this._scene.canvas.height

        pxX *= elWidth
        pxY *= elHeight

        return {
            x: pxX - (elWidth / 2),
            // y-coord is inverted (positive up in world space, positive down in
            // pixel space)
            y: -pxY + (elHeight / 2),
            z: z,
        };

    },

    unproject(x, y) {

        // The 3D world size of the entity
        const {
            width: elWidth,
            height: elHeight
        } = this.el.components.geometry.data;

        const pxX = (x + (elWidth / 2));
        // y-coord is inverted (positive up in world space, positive down in
        // pixel space)
        const pxY = ((elHeight / 2) - y);


        var nx = pxX / elWidth
        var ny = pxY / elHeight

        nx *= this._scene.canvas.width
        ny *= this._scene.canvas.height

        // Return the lat / long of that pixel on the map
        var latLng = this._mapInstance.layerPointToLatLng([nx, ny])
        return {
            lat: latLng.lat,
            lon: latLng.lng
        }
    },
    distanceTo(lngLat) {
        // TODO add altitude, has it an effect?
        return L.latLng(lngLat).distanceTo(this.data.center)
    },
    altitudeTo(lngLat) {

        var data = this.data

        const {
            x: currX,
            y: currY
        } = this._mapInstance.latLngToLayerPoint([data.center[1], data.center[0]]);

        var currAltitude = this._getAltitudeFromXY(currX, currY)

        return currAltitude - this.getAltitude(lngLat)
    },
    getAltitude(lngLat) {
        const {
            x: givenX,
            y: givenY
        } = this._mapInstance.latLngToLayerPoint([lngLat[1], lngLat[0]]);

        return this._getAltitudeFromXY(givenX, givenY)
    }
});